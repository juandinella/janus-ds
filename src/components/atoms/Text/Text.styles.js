import { css } from 'styled-components'
import layout from '../../../styles/theme/layout'

const getSizeForMobile = (size) => {
  if (Array.isArray(size) && size[0]) {
    return `font-size: var(--font-size-${size[0]});`
  }
  if (typeof size === 'string') {
    return `font-size: var(--font-size-${size});`
  }
  return ''
}

const getSizeForDesktop = (size) => {
  if (Array.isArray(size) && size[1]) {
    return `
      @media screen and (min-width: ${layout.breakpoints.sm}) {
        font-size: var(--font-size-${size[1]});
      }
    `
  }
  return ''
}

export default css`
  display: flex;
  line-height: 1.25;

  ${(props) => css`
    ${getSizeForMobile(props.size)}
    ${getSizeForDesktop(props.size)}
  `}

  ${(props) =>
    props.weight &&
    css`
      font-weight: var(--font-weight-${(props) => props.weight});
    `}

  ${(props) =>
    props.alignment &&
    css`
      text-align: ${(props) => props.alignment};
    `}

  ${(props) =>
    props.color &&
    css`
      color: var(--color-neutral-${(props) => props.color});
    `}

  ${(props) =>
    props.transform &&
    css`
      text-transform: ${(props) => props.transform};
    `}

  ${(props) =>
    props.lineClamp &&
    css`
      display: -webkit-box;
      line-clamp: ${(props) => props.lineClamp};
      -webkit-line-clamp: ${(props) => props.lineClamp};
      -webkit-box-orient: vertical;
      overflow: hidden;
    `}
`
