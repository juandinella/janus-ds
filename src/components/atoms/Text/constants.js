export const options = {
  sizes: ['xs', 'sm', 'md', 'lg', 'xl'],
  types: ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'span', 'p'],
  weights: ['light', 'regular', 'medium', 'bold'],
  alignments: ['left', 'center', 'right'],
  transforms: ['capitalize', 'uppercase', 'lowercase', 'initial'],
  colors: [
    '50',
    '100',
    '200',
    '300',
    '400',
    '500',
    '600',
    '700',
    '800',
    '900',
    '950',
  ],
}
