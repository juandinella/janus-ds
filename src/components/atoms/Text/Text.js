import isPropValid from '@emotion/is-prop-valid'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import { options } from './constants'
import styles from './Text.styles'

/**
 * Text
 */

const shouldForwardProp = (prop) => isPropValid(prop)
const StyledText = styled.p.withConfig({shouldForwardProp})`${styles}`

const Text = ({ id, type, size, weight, alignment, transform, color, lineClamp, children}) => {
  return <StyledText id={id} as={type} size={size} weight={weight} alignment={alignment} transform={transform} lineClamp={lineClamp} color={color}>{children}</StyledText>
}

Text.displayName = 'Text'

Text.propTypes = {
  /** Component id */
  id: PropTypes.string,
  /** Content to be included in the component */
  children: PropTypes.node.isRequired,
  /* One of: h1, h2, h3, h4, h5, h6, span or p */
  type: PropTypes.oneOf(options.types).isRequired,
  /** Size of the text */
  size: PropTypes.oneOfType([
    PropTypes.oneOf(options.sizes),
    PropTypes.arrayOf(PropTypes.oneOf(options.sizes))
  ]),
  /** Text's font weight */
  weight: PropTypes.oneOf(options.weights),
  /** Transform text */
  transform: PropTypes.oneOf(options.transforms),
  /** Text's Alignmnet */
  alignment: PropTypes.oneOf(options.alignments),
  /** Text's color */
  color: PropTypes.oneOf(options.colors),
  /** Clamp text */
  lineClamp: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
}

Text.defaultProps = {
  type: 'p',
  color: 'neutral-darker',
}

export default Text
