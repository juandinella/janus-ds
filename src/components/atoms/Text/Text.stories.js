import { options } from './constants'
import Text from './Text'

import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(Text)
const ListTemplate = getListTemplate(Text)

export default {
  title: 'Atoms/Text',
  component: Text,
  args: {
    children:
      "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    size: 'lg',
  },
  argTypes: {
    ...getOptionsArgTypes(options.sizes, 'size'),
    ...getOptionsArgTypes(options.types, 'type'),
    ...getOptionsArgTypes(options.weights, 'weight'),
    ...getOptionsArgTypes(options.alignments, 'alignment'),
    ...getOptionsArgTypes(options.transforms, 'transform'),
    ...getOptionsArgTypes(options.colors, 'color'),
  },
}

export const Default = Template.bind({})

export const Responsive = { args: { size: ['xs', 'xl'] } }

export const LineClamp = { args: { lineClamp: 1 } }

export const Sizes = ListTemplate.bind({})
Sizes.args = { items: options.sizes.map((size) => ({ size })) }

export const Weights = ListTemplate.bind({})
Weights.args = { items: options.weights.map((weight) => ({ weight })) }

export const Alignments = ListTemplate.bind({})
Alignments.args = {
  items: options.alignments.map((alignment) => ({ alignment })),
}

export const Transforms = ListTemplate.bind({})
Transforms.args = {
  items: options.transforms.map((transform) => ({ transform })),
}

export const Colors = ListTemplate.bind({})
Colors.args = { items: options.colors.map((color) => ({ color })) }
