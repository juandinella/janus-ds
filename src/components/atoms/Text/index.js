import Text from './Text'

export default Text
export { options } from './constants'
