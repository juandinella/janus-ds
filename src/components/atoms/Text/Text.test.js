import initStoryshots from '@storybook/addon-storyshots'
import { render, screen, cleanup } from '@testing-library/react'
import React from 'react'
import Text from './Text'

initStoryshots({
  storyKindRegex: /^Atoms\/Text$/,
})

const componentProps = {
  id: 'unique_id',
  children: 'Go!',
}

afterEach(cleanup)

describe('<Text />', () => {
  test('should render component correctly', () => {
    render(<Text {...componentProps} />)
    const text = screen.getByTestId(componentProps.id)
    expect(text).toBeInTheDocument()
  })

  test('should render content correctly', () => {
    render(<Text {...componentProps} />)
    const p = screen.getByText(componentProps.children)
    expect(p).toBeInTheDocument()
  })

  test('should render type correctly', () => {
    render(<Text {...componentProps} type="h4" />)
    const h4 = screen.getByRole('heading', { level: 4 })
    expect(h4).toBeInTheDocument()
  })
})
