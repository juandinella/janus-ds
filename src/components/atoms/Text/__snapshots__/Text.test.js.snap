// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Text Alignments 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <md
    class="sc-81qx22-0 hcZLQm"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 dqLZli"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 lnJIIx"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
</div>
`;

exports[`Storyshots Atoms/Text Colors 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <md
    class="sc-81qx22-0 fLSNtZ"
    color="50"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 fLjdUJ"
    color="100"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 dgDhWy"
    color="200"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 cYhxKP"
    color="300"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 jfByGo"
    color="400"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 iXfOuF"
    color="500"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 gszSwu"
    color="600"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 gkeikL"
    color="700"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 dHeLvU"
    color="800"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 dyJbkl"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 dhxJTs"
    color="950"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
</div>
`;

exports[`Storyshots Atoms/Text Default 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <md
    class="sc-81qx22-0 dyJbkl"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
</div>
`;

exports[`Storyshots Atoms/Text Line Clamp 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <md
    class="sc-81qx22-0 iOJaPO"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
</div>
`;

exports[`Storyshots Atoms/Text Responsive 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <md
    class="sc-81qx22-0 faAHlJ"
    color="900"
    size={
      [
        "xs",
        "xl",
      ]
    }
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
</div>
`;

exports[`Storyshots Atoms/Text Sizes 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <md
    class="sc-81qx22-0 bHuaYz"
    color="900"
    size="xs"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 lmIqDq"
    color="900"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 dyJbkl"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 gXnhrD"
    color="900"
    size="lg"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 cqChhc"
    color="900"
    size="xl"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
</div>
`;

exports[`Storyshots Atoms/Text Transforms 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <md
    class="sc-81qx22-0 ghSDQs"
    color="900"
    size="md"
    transform="capitalize"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 eANyqi"
    color="900"
    size="md"
    transform="uppercase"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 ipviqP"
    color="900"
    size="md"
    transform="lowercase"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 jussMO"
    color="900"
    size="md"
    transform="initial"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
</div>
`;

exports[`Storyshots Atoms/Text Weights 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <md
    class="sc-81qx22-0 hFJsOv"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 kmQFzn"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 cRqnnI"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
  <md
    class="sc-81qx22-0 fjdNXQ"
    color="900"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </md>
</div>
`;
